presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java

```diff
@@ -210,13 +239,67 @@ public void addInput(Page page)
 
     private void handleTooLargePredicate()
     {
-        // The resulting predicate is too large, allow all probe-side values to be read.
-        dynamicPredicateConsumer.accept(TupleDomain.all());
+        // The resulting predicate is too large
+        if (minMaxValues) {
+            // convert to min/max per column for orderable types
+            for (Integer channelIndex : orderableChannels) {
+                Block block = blockBuilders[channelIndex].build();
+                updateMinMaxValues(block, channelIndex);
+            }
+        }
+        else {
+            // allow all probe-side values to be read.
+            dynamicPredicateConsumer.accept(TupleDomain.all());
+        }
         // Drop references to collected values.
         valueSets = null;
         blockBuilders = null;
     }
 
+    private void updateMinMaxValues(Block block, int channelIndex)
+    {
+        Type type = channels.get(channelIndex).type;
+        int minValuePosition = -1;
+        int maxValuePosition = -1;
+
+        for (int position = 0; position < block.getPositionCount(); ++position) {
+            if (block.isNull(position)) {
+                continue;
+            }
+            if (minValuePosition == -1) {
+                // First non-null value
+                minValuePosition = position;
+                maxValuePosition = position;
+            }
+            else {
+                if (type.compareTo(block, position, block, minValuePosition) < 0) {
+                    minValuePosition = position;
+                }
+                else if (type.compareTo(block, position, block, maxValuePosition) > 0) {
+                    maxValuePosition = position;
+                }
+            }
```

Comment:
@sopel39 This implementation can run into https://github.com/prestosql/presto/issues/4272
`TypeUtils#isFloatingPointNaN` currently works on Object, so we would end up converting every position to an Object using `readNativeValue` if we try to add that check.
I think we could either skip DOUBLE and REAL types from this optimization or we could look into adding `boolean Block#isFloatingPointNaN(int position)`. WDYT ?

Reply:
For now I've skipped DOUBLE and REAL types

Reply:
For nested types there is no failure in the query as in #4272
We are able to create TupleDomain out of an array of Double.NaN as `isFloatingPointNaN` does not look into nested types. So I think for nested types, this change probably doesn't make the situation any worse.